### DISCUSSION TITLE
SWIG Python libavrdude wrapper

### DISCUSSION THREAD

**Discussion by dl8dtl** [avrdudes/avrdude#1715] [Ideas]
I just opened a WIP (work in progress) pull request for a SWIG Python wrapper for libavrdude.

I've been toying with that idea all the way back when we started segregating libavrdude from the actual frontend code, years ago. However, I never found enough time to really materialize the idea in any way. Now that there are so many nice developers driving the progress of AVRDUDE (thanks!), I could lean back a bit, and get back to that idea.

I confess: I imagined it would be easier … there's a lot of legwork needed, to wrap everything that would be useful from the Python level. Fortunately, a lot of things are basically opaque stuff that is only required inside libavrdude itself, so it does not need to be wrapped.

Everything is now in a state were I would consider it to be "hello world": config files can be read, the list of AVR devices can be walked, and for each device, the possible memory regions can be examined.

I think SWIG might also benefit a lot from turning as many `#define`s into `enum` literals as possible, as they are typed, and can therefor be wrapped automatically.

Any suggestions or even code contributions (to my branch) are welcome!

**Comment by mcuee** [comment]
@dl8dtl 

I think a Python binding for libavrdude is good to have.

My question is why using SWIG? The more popular ways now seem to be using ctypes or cython. Example from the projects I am involved in. From my testing, libftdi's built-in SWIG based binding is the most problematic.

pyusb -- ctypes
python-libusb1 -- ctypes
cython-hidapi -- cython
libftdi built-in python binding -- SWIG
pylibftdi -- ctypes

**Reply by dl8dtl** [reply]
Well, frankly, because I don't have even less of an idea about cython or ctypes. ;-)
Anyway, what I've been learning during the few days I've been dealing with it, libavrdude exposes many internal things which are completely irrelevant to a higher-level program. Everything read from the config file(s) must be treated read-only by the program, so most of the fields we are parsing from the config file(s) are completely uninteresting to the higher layers.
If it doesn't cost too much effort, could you sketch a short snippet how a ctypes binding would be started?

**Reply by mcuee** [reply]
I actually do not know much about ctypes and even python, but here is a simple introduction.
https://docs.python.org/3/library/ctypes.html

I think the following shows how simple it can be.
https://github.com/apmorton/pyhidapi/blob/master/hid/__init__.py

Cython is probably a bit more complicated.
https://github.com/trezor/cython-hidapi/blob/master/hid.pyx

**Reply by mcuee** [reply]
@dl8dtl 

The following is based on my own experiences building various Python bindings under different OS and toolchain.
I am more for ctypes as it has never given me problems building under different OS (eg: pyusb, python-libusb1, etc)
cython sometimes gives a bit of issue but mostly okay (cython-hidapi).
SWIG is always a bit problematic (libftdi built-in python binding) espcially under Window.

**Comment by mcuee** [comment]
@ndim 

Not so sure if you have some inputs here as well. Thanks.

**Reply by ndim** [reply]
I do have some ideas, but TBH, the time I can spend on avrdude is less than all the things I have ideas for would take to discuss and develop right now.

I had therefore hoped to kick the API can down the road for a bit longer until I have concluded most of the build stuff, but I fear I have to take a look into API design next after all and reschedule the build stuff for later.

**Reply by mcuee** [reply]
For the build stuff, there are things related to libavrdude. Before the ABI/API discussions conclude, it is probably better that avrdude uses static link with libavrdude. So at least some of the build improvement activities related to libavrdude can be delayed.

**Comment by mcuee** [comment]
> Anyway, what I've been learning during the few days I've been dealing with it, libavrdude exposes many internal things which are completely irrelevant to a higher-level program. Everything read from the config file(s) must be treated read-only by the program, so most of the fields we are parsing from the config file(s) are completely uninteresting to the higher layers.

@dl8dtl 

The comments by @ndim here may be related. Maybe now it is time to clean up libavrdude, probably @stefanrueger, @ndim and you can help on this front.
https://github.com/avrdudes/avrdude/issues/1159#issuecomment-1975268173

>The loooong list of part specific constants macro definitions might need to be converted into data structures and functions acting on those which represent the semantics of those compile time definitions at runtime. So the API/ABI would only cover a few functions getting and dealing with those part specific data structures which are provided by libavrdude. Then those data structures and functions could remain stable in the longer term. You might need to add a new function from time to time which does not break existing API or ABI.

**Comment by stefanrueger** [comment]
> I had therefore hoped to kick the API can down the road for a bit longer until I have concluded most of the build stuff

I still think it is OK to fix the build system first and look at the API/ABI later.

>The loooong list of part specific constants macro definitions might need to be converted into data structures and functions acting on those which represent the semantics of those compile time definitions at runtime. So the API/ABI would only cover a few functions getting and dealing with those part specific data structures which are provided by libavrdude.

If that means encapsulating some of the properties of parts by the use of get_...() functions and hiding access to components this is doable. If the end-goal is to not expose the internal structure of PROGRAMMER, AVRPART or AVRMEM in the API, that's also doable. This kind of means replacing `p->property` with `get_property(p)`.

What has happened in the past is that, for example, a new part came out with an "unusual" property that the page size for erasing is 4 times bigger than the page size for writing. AVRDUDE needed to find a way to describe a part as having this property and *extended* the AVRPART structure by a component `n_page_erase` that is set when reading in the `avrdude.conf.in` file. That change would break the ABI b/c the structure size changes. As new parts and new programmers come out and as AVRDUDE got better at closely modelling parts and programmers, new features or components would be added to these structures. Of course, we could hide a lot of this by adding 20 new ints `extension0`, ... `extension19` to the main structures once, and "allocate" a new feature by *renaming*, say, `extension0` to `n_page_erase`  internally. Internally, because we would no longer expose the structure components in `libavrdude.h`. So, the ABI could be stable as long as we don't  run out of these spare bytes in the structures, and the API would change less dramatically by adding new functions.

But, and this is the crucial bit here, this should by and large be independent of the build process. So, I'd argue to get this [the build system] right first.

**Reply by mcuee** [reply]
@stefanrueger 

Instead of adding `extension0, extension, ..., extension19`, can we use C variadic in the API?

But I am not so sure if there is an issue using variadic for Python binding or not.
Reference discussion: 
* https://github.com/vpelletier/python-libusb1/issues/75

**Reply by ndim** [reply]
A libavrdude library user should not need to know the size or the the memory layout of libavrdude internal structures like PROGRAMMER or AVRPART or AVRMEM. A libavrdude user never needs to copy or free such a structure, and accesses to struct members can be done by functions with a stable function signature.

As I understand it, libavrdude reads in the `avrdude.conf` file once and builds arrays of data structures with some pointers to each other. All that data can probably be considered readonly data after that step has finished even inside libavrdude, and especially from the POV of a library user.

So libavrdude can change the size and memory layout of its data structures as it likes without breaking API or ABI. And if a function needs to return a string from one of those data structures, it does not even need to involve `strncpy()` and `free()` on the library user's side, because such a function can just return a `const char *` pointing to data which actually does not change.

A very basic example for a fictitious "libad" library and a utility using "libad", which I have written without taking a look at libavrdude internals at all:

```C
/* libad.h */
typedef struct ad_programmer ad_programmer;
typedef struct ad_part ad_part;
typedef struct ad_context ad_context;

ad_context *ad_context_new(void);
void ad_context_free(ad_context *ctx);

int ad_write(ad_context *ctx, ad_programmer *prog, ad_part *part, void *buf, size_t sz);
```

```C
/* libad.c */
#include <stddef.h>
#include <libad.h>

struct ad_part {
  ...
  size_t n_page_erase;
  ...
};

struct ad_programmer {
  ...
};

int ad_write(ad_context *ctx, ad_programmer *prog, ad_part *part, void *buf, size_t sz)
{
   ctx->errno = 0;
   ...
   ... deal with n_page_erase ...
   ...
}

struct ad_context {
  ad_programmer *programmers;
  ad_parts *parts;
};

ad_context *ad_context_new(void)
{
  ad_context *ctx = calloc(1, sizeof(*ctx));
  ... parse avrdude.conf into ctx->programmers and ctx->parts and ...
  return ctx;  
}

void ad_context_free(ad_context *ctx)
{
  ... free members of ctx ...
  free(ctx);
}
```

```C
/* main.c */
#include <libad.h>

int main(void)
{
  ad_context *ctx = ad_context_new();
  ad_programmer *prog = ad_...(ctx);
  ad_part *part = ad_...(ctx);
  uint8_t buf[256];
  if (ad_write(ctx, prog, part, buf, sizeof(buf))) {
    /* deal with ad_write errors */
    fprintf(stderr, "ad error %d\n", ctx->errno);
  }
  ad_context_free(ctx);
  return 0;
}
```

The library user `main.c` only sees the API from `libad.h`, which only uses pointers to incomplete types, i.e. an opaque type.

As `main.c` never sees the size or the memory layout of any of the `ad_*` data structures, it can never allocate such a structure, perform pointer `++` or similar on a pointer to one, or access structure members, and therefore `libad.c` can chose to change the structure size and layout and add and remove struct members without ever breaking the API or ABI.

It should be a not insignificant amount of not very complex code to adapt such data structures and the functions operating on them to a Python or C++ or ... API wrapper.

**Reply by dl8dtl** [reply]
The disadvantage is that it requires drafting a completely new library – without anyone known to be willing to use it right now.

libavrdude was a simple quick way to just segregate the existing pieces of code into a library backend, and a UI (CLI) frontend. Even that has proven to already cause quite a bit of extra work – still, with no known users (yet).

Basically, what you are describing is a bit of that what the existing pieces of my SWIG wrapper are doing, just that you don't have a single function for each any any possible access to struct fields, but instead wrap up what I believed to be relevant fields into a Python dict. That does not require spending more work in to the backend library, only into the interface definition.

**Reply by ndim** [reply]
Yes, that would be a completely new library, at least the parts directed to the outside world via its API and ABI. The programmer protocol code etc. would not be completely new. Still a lot of work.

The existing libavrdude API shows that it was not particularly designed ("simple quick way of segregating existing pieces of code" is exactly what the API looks like), and that is why I would say it is unsuitable for building at least shared libraries of. Static libraries might be helpful for some scenarios.

If I needed to incorporate parts of the `avrdude` capabilities into some code of mine, I would probably choose spawning an `avrdude` process over linking to `libavrdude`, just for the more stable API.

FYI, a few weeks ago in one of the issues I was active in, there was a library user... here: https://github.com/avrdudes/avrdude/issues/1683#issuecomment-1937818082

I am unfamiliar with this space, but there might be projects which have been spawning `avrdude` processes in the past and might be interested to switching to a library.

As to the functions... you would need iteration functions over programmers and parts, lookup functions for ports, programmers, and parts, and many more. I just wanted to demonstrate how incomplete types can be used in C to allow a library to internally change its data structures without necessarily breaking API or even ABI.

**Reply by mcuee** [reply]
Interestingly Homebrew installs dynamic linking avrdude binary and static linked `libavrdude.a` as of now.
https://github.com/Homebrew/homebrew-core/blob/589f58b81e8d3a03aebcf46d89821faef84ed040/Formula/a/avrdude.rb

```
  def install
    args = std_cmake_args + ["-DCMAKE_INSTALL_SYSCONFDIR=#{etc}"]
    shared_args = ["-DBUILD_SHARED_LIBS=ON", "-DCMAKE_INSTALL_RPATH=#{rpath}"]
    shared_args << "-DCMAKE_SHARED_LINKER_FLAGS=-Wl,-undefined,dynamic_lookup" if OS.mac?

    system "cmake", "-S", ".", "-B", "build/shared", *args, *shared_args
    system "cmake", "--build", "build/shared"
    system "cmake", "--install", "build/shared"

    system "cmake", "-S", ".", "-B", "build/static", *args
    system "cmake", "--build", "build/static"
    lib.install "build/static/src/libavrdude.a"
  end
```

**Reply by mcuee** [reply]
Without the trick of using `-DCMAKE_SHARED_LINKER_FLAGS=-Wl,-undefined,dynamic_lookup`, macOS shared libavrdude library build will fail. But I am not so sure if that build flag is correct or not.

* https://github.com/avrdudes/avrdude/issues/1414

**Comment by dl8dtl** [comment]
> The existing libavrdude API shows that it was not particularly designed ("simple quick way of segregating existing pieces of code" is exactly what the API looks like), and that is why I would say it is unsuitable for building at least shared libraries of. Static libraries might be helpful for some scenarios.

Actually, the SWIG wrapper (as I have it now) isn't using the library itself, but it recompiles everything statically – in order to eventually supply a shared lib for Python. I think it might even be able to use the existing shared lib to build upon though, it's just that the CMake stuff was easier for me to do that way (I'm absolutely unfamilar with CMake).

> If I needed to incorporate parts of the `avrdude` capabilities into some code of mine, I would probably choose spawning an `avrdude` process over linking to `libavrdude`, just for the more stable API.

That's what many others did in the past (and still do), but I think it's really a crock rather than a solution. You'd have to parse all the messages shown, and unless you implement a kind of pseudo TTY wrapper (as e.g. Arduino does), there's no chance for things like a progress indication that really integrates into that wrapper.

All that once eventually got us to the idea of segregating frontend and backend in a clearer manner than it used to be before.

> As to the functions... you would need iteration functions over programmers and parts, lookup functions for ports, programmers, and parts, and many more. I just wanted to demonstrate how incomplete types can be used in C to allow a library to internally change its data structures without necessarily breaking API or even ABI.

The SWIG wrapper just wraps the list access functions (which came from Brian Dean, and have always been there since the start of AVRDUDE), and just offers a few helper functions (basically typecast), to access things like programmer, parts, or memory regions. That is kind of a compromise between effort (for the wrapper) and not having to add anything to the backend at all.

The completely new abstraction library is certainly a good idea, but I wouldn't want to do it without a potential "customer" since I assume it is really quite some effort (and then still requires [some] additional effort for non-C wrappers).

**Reply by mcuee** [reply]
> Actually, the SWIG wrapper (as I have it now) isn't using the library itself, but it recompiles everything statically – in order to eventually supply a shared lib for Python. I think it might even be able to use the existing shared lib to build upon though, it's just that the CMake stuff was easier for me to do that way (I'm absolutely unfamilar with CMake).

Interesting approach and I kind of understand why you use SWIG now.

However, SWIG based Python bindings are more difficult to build cross-platform based on my own experiences. We can already see that there are problems with macOS (should be able to be sorted out as only one step is mssing) and Windows (more difficult to get right for both MSYS2 mingw build and MSVC build).
* https://github.com/avrdudes/avrdude/pull/1714

**Comment by mcuee** [comment]
> Yes, that would be a completely new library, at least the parts directed to the outside world via its API and ABI. The programmer protocol code etc. would not be completely new. Still a lot of work.
> ...
> I am unfamiliar with this space, but there might be projects which have been spawning `avrdude` processes in the past and might be interested to switching to a library.
> ...
> As to the functions... you would need iteration functions over programmers and parts, lookup functions for ports, programmers, and parts, and many more. I just wanted to demonstrate how incomplete types can be used in C to allow a library to internally change its data structures without necessarily breaking API or even ABI.

Another thing, there could be a licensing concerns as well.

For those choosing to spawning the `avrdude` process, technically they can use different license from arvdude (GPL).

For example, AVRDUESS is using GPL. But technically different license can be used.
https://github.com/ZakKemble/AVRDUDESS

As of now, libavrdude will be covered by GPL without linking exception. Technically speaking, LGPL will be a more suitable license if we want to allow more users to use it.

A new library may be able to use LGPL (or GPL with linking exception).

As of now, this is a moot point though as there are not well known projects using libarvdude.

**Reply by dl8dtl** [reply]
Naturally, with my BSD roots, I (and also Brian Dean) would have favored a BSD license. It was Ted Roth by that time who convinced Brian to put the project under GPL. I think it can just remain there, including derivatives. So far, I don't know of anyone who'd have complained about that.

**Comment by dl8dtl** [comment]
> I think SWIG might also benefit a lot from turning as many `#define`s into `enum` literals as possible, as they are typed, and can therefor be wrapped automatically.

This commit is a typical example:

https://github.com/dl8dtl/avrdude/commit/377d3f353c6834e051d84edd3f5eb33df6c7db98

Using the `enum`, it can be simply copy&pasted into the interface definitions, so people could use them as `ad.PM_ISP` and so on.

**Comment by dl8dtl** [comment]
Btw., just for a test, I turned the CMake library computation into a shared lib, and I could successfully use that shared lib in the SWIG wrapper from Python. I don't think this is really the way to go right now, but it works.

If someone is interested:
```
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 95223930..a39e766e 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -286,6 +286,7 @@ set(SOURCES
     )
 
 add_library(libavrdude
+    SHARED
     ${SOURCES}
     )
 
@@ -341,7 +342,7 @@ endif()
 
 if(HAVE_SWIG)
   include (UseSWIG)
-  swig_add_library(swig_avrdude LANGUAGE Python SOURCES libavrdude.i ${SOURCES})
+  swig_add_library(swig_avrdude LANGUAGE Python SOURCES libavrdude.i)
   target_include_directories(swig_avrdude
     PUBLIC
     "${PYTHON_INCLUDE_PATH}"
@@ -351,8 +352,10 @@ if(HAVE_SWIG)
     "${LIBUSB_COMPAT_DIR}"
     ${EXTRA_WINDOWS_INCLUDES}
   )
+  target_link_directories(swig_avrdude PUBLIC "${PROJECT_BINARY_DIR}/src")
   target_link_libraries(swig_avrdude
     PRIVATE
+    libavrdude.so
     ${LIB_MATH}
     ${LIB_LIBELF}
     ${LIB_LIBUSB}
```

**Comment by mcuee** [comment]
This has been part of avrdude 8.0 release.
